### PR TITLE
feat(mission_planner): add mrm route planner

### PR DESCRIPTION
**Pull request by purewater0901**
## Description
Add MRM reroute implementation in the mission planner. If there is an MRM manager, and it requires rerouting in an appropriate situation, the mission planner creates a new route for the MRM manager. It also sends the MRM route to the behavior path planner. Note that if the reroute success, it only published mrm route via mrm_route_pub_.

## Tests performed

Not applicable.

## Effects on system behavior

If it's under MRM state, it can create an MRM route.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 1]
```diff
@@ -367,6 +488,11 @@ void MissionPlanner::on_modified_goal(const ModifiedGoal::Message::ConstSharedPt
     RCLCPP_ERROR(get_logger(), "Normal route has not set yet.");
     return;
   }
+  if (mrm_route_) {
```
Comment (isamu-takagi) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:491]: Isn't modified_goal necessary even in MRM? There may be obstacles at the goal of the mrm_route.

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 2]
```diff
@@ -334,19 +367,107 @@ void MissionPlanner::on_set_mrm_route(
...
 void MissionPlanner::on_clear_mrm_route(
-  const ClearMrmRoute::Service::Request::SharedPtr req,
+  [[maybe_unused]] const ClearMrmRoute::Service::Request::SharedPtr req,
   const ClearMrmRoute::Service::Response::SharedPtr res)
 {
-  // TODO(Yutaka Shimizu): reroute for MRM
-  (void)req;
-  (void)res;
+  using ResponseCode = autoware_adapi_v1_msgs::srv::SetRoutePoints::Response;
+  change_state(RouteState::Message::CHANGING);
```
Comment (isamu-takagi) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:419]: change_state is done after all guard conditions.

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 3]
```diff
@@ -468,6 +594,11 @@ void MissionPlanner::on_change_route_points(
     throw component_interface_utils::ServiceException(
       ResponseCode::ERROR_PLANNER_UNREADY, "Normal route is not set.");
   }
+  if (mrm_route_) {
+    RCLCPP_ERROR_THROTTLE(
+      get_logger(), *get_clock(), 5000, "Cannot reroute in the emergency state");
+    return;
```
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:600]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/f67aa65a76024ea55319db16ad1dbf6e02714f41

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 4]
```diff
@@ -334,19 +367,107 @@ void MissionPlanner::on_set_mrm_route(
...
+    change_route(*normal_route_);
+  }
+  change_state(RouteState::Message::SET);
+  res->status.success = false;
 }
 
 // NOTE: The route interface should be mutually exclusive by callback group.
 void MissionPlanner::on_clear_mrm_route(
-  const ClearMrmRoute::Service::Request::SharedPtr req,
+  [[maybe_unused]] const ClearMrmRoute::Service::Request::SharedPtr req,
```
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/f67aa65a76024ea55319db16ad1dbf6e02714f41

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 5]
```diff
@@ -313,6 +341,11 @@ void MissionPlanner::on_set_route_points(
     throw component_interface_utils::ServiceException(
       ResponseCode::ERROR_PLANNER_UNREADY, "The vehicle pose is not received.");
   }
+  if (mrm_route_) {
+    RCLCPP_ERROR_THROTTLE(
+      get_logger(), *get_clock(), 5000, "Cannot set route in the emergency state");
+    return;
```
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:347]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/f67aa65a76024ea55319db16ad1dbf6e02714f41

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 6]
```diff
@@ -278,6 +301,11 @@ void MissionPlanner::on_set_route(
...
```
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:307]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/f67aa65a76024ea55319db16ad1dbf6e02714f41

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 2 above)
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/41357dc9a761fabd82eeb1346a74baad0bbc89c9

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 7]
```diff
@@ -468,6 +594,11 @@ void MissionPlanner::on_change_route_points(
     throw component_interface_utils::ServiceException(
       ResponseCode::ERROR_PLANNER_UNREADY, "Normal route is not set.");
   }
+  if (mrm_route_) {
```
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:597]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/41357dc9a761fabd82eeb1346a74baad0bbc89c9

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 8]
```diff
@@ -334,19 +367,107 @@ void MissionPlanner::on_set_mrm_route(
...
+      ResponseCode::ERROR_PLANNER_UNREADY, "The vehicle pose is not received.");
+  }
+  if (!normal_route_) {
+    change_state(RouteState::Message::SET);
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "Normal route is not set.");
+  }
+  if (!mrm_route_) {
+    change_state(RouteState::Message::SET);
+    throw component_interface_utils::ServiceException(
```
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/41357dc9a761fabd82eeb1346a74baad0bbc89c9

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 1 above)
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:486]: I'll submit a new PR for handling modified goal during the MRM state next week, I don't want to change goals in the emergency state

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 9]
```diff
@@ -334,19 +367,107 @@ void MissionPlanner::on_set_mrm_route(
...
+  }
+
+  // Failed to reroute
+  if (mrm_route_) {
+    // if it has the old mrm route, use it
+    change_mrm_route(*mrm_route_);
+  } else {
+    // failed to go to mrm state, use normal route
+    change_route(*normal_route_);
+  }
```
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:406]: If mrm module asks the mission planner to change route when it's under the mrm state, the old route should be previous mrm_route.

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 10]
```diff
@@ -334,19 +367,107 @@ void MissionPlanner::on_set_mrm_route(
...
+  if (!odometry_) {
+    change_state(RouteState::Message::SET);
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "The vehicle pose is not received.");
+  }
+  if (!normal_route_) {
+    change_state(RouteState::Message::SET);
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "Normal route is not set.");
+  }
```
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/df054e151b1dfd3abe2727c2204be8a9e3c47cf8

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 11]
```diff
@@ -334,19 +367,107 @@ void MissionPlanner::on_set_mrm_route(
...
```
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/df054e151b1dfd3abe2727c2204be8a9e3c47cf8

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 1 above)
Comment (isamu-takagi) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:501]: I see.

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp]: Fix with https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/698cffa98c17ba08bc100dfc921b70223180bcef

[the remark this answers is not in the document]
Comment (isamu-takagi) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp]: > Does create_route have any side effects?

Note:  `planner.plan` in `create_route` updates planner local variable, so it should be reset by `planner.updateRoute` in `change_route/change_mrm_route`.

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 12]
```diff
@@ -334,19 +365,124 @@ void MissionPlanner::on_set_mrm_route(
...
+      ResponseCode::ERROR_PLANNER_FAILED, "Failed to plan a new route.");
+  }
+
+  if (!normal_route_) {
+    // if it does not set normal route, just use the new planned route
+    change_mrm_route(new_route);
+    change_state(RouteState::Message::SET);
+    res->status.success = true;
+    return;
+  }
```
Comment (isamu-takagi) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:397]: This should be moved between step1 and step2. Safety is not checked when only mrm_route is set.

Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp]: done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/d16313cb6f9ffa9065fadd59f5f30561b91f5d93

**Review by isamu-takagi — APPROVED** [review]
LGTM

**Pull request merged**